{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f39793c3-e3f9-4dd1-baaf-bf82443987e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CondaValueError: prefix already exists: /home/kbanerj1/.conda/envs/JupyterLab\n",
      "\n",
      "Requirement already satisfied: ipykernel in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (6.29.5)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (0.2.2)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (1.8.5)\n",
      "Requirement already satisfied: ipython>=7.23.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (8.18.1)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (8.6.2)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (5.7.2)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (0.1.7)\n",
      "Requirement already satisfied: nest-asyncio in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (1.6.0)\n",
      "Requirement already satisfied: packaging in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (24.1)\n",
      "Requirement already satisfied: psutil in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (6.0.0)\n",
      "Requirement already satisfied: pyzmq>=24 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (26.2.0)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (6.4.1)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (5.14.3)\n",
      "Requirement already satisfied: decorator in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (0.19.1)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (3.0.47)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (2.18.0)\n",
      "Requirement already satisfied: stack-data in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (0.6.3)\n",
      "Requirement already satisfied: typing-extensions in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (4.12.2)\n",
      "Requirement already satisfied: exceptiongroup in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (1.2.2)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (4.9.0)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-client>=6.1.12->ipykernel) (8.5.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.9.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (4.3.3)\n",
      "Requirement already satisfied: zipp>=3.20 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=6.1.12->ipykernel) (3.20.2)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel) (0.8.4)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from pexpect>4.3->ipython>=7.23.1->ipykernel) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from prompt-toolkit<3.1.0,>=3.0.41->ipython>=7.23.1->ipykernel) (0.2.13)\n",
      "Requirement already satisfied: six>=1.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.16.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.1.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (0.2.3)\n",
      "Installed kernelspec JupyterLab in /home/kbanerj1/.local/share/jupyter/kernels/jupyterlab\n"
     ]
    }
   ],
   "source": [
    "!conda env create -f environment.yml\n",
    "!pip install ipykernel\n",
    "!python -m ipykernel install --user --name JupyterLab --display-name \"Python (JupyterLab)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "236bbba3-8faf-443a-a59f-0bf2f99fa357",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fff16a37-85d6-4a0f-8231-7afce4a9aa7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Current Directory:\", os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77c57fe-45cd-4a91-b4e9-415f8298a08d",
   "metadata": {},
   "source": [
    "## Reading Gene Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9b533363-49c6-4a8e-8192-b4f1ff3441af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gene_df(filepath, filename):\n",
    "    '''\n",
    "    Function that reads a gene file and processes the gene data with the file name.\n",
    "    Skips rows with unwanted gene names (e.g., N_unmapped, N_multimapping, N_noFeature, N_ambiguous).\n",
    "    Retains only the tpm_unstranded column.\n",
    "    '''\n",
    "    \n",
    "    df1 = pd.read_csv(filepath, sep='\\t', skiprows=1)\n",
    "    unwanted_genes = ['N_unmapped', 'N_multimapping', 'N_noFeature', 'N_ambiguous']\n",
    "    df1 = df1[~df1['gene_id'].isin(unwanted_genes)]\n",
    "    df1 = df1[['gene_name', 'tpm_unstranded']]\n",
    "    df1.set_index('gene_name', inplace=True)\n",
    "    df1 = df1.T\n",
    "    processed_filename = filename.split('.')[0]\n",
    "    df1.index = [processed_filename]\n",
    "    df1 = df1.reset_index()\n",
    "    df1.rename(columns={'index': 'file_name'}, inplace=True)\n",
    "    \n",
    "    return df1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c75ff3e",
   "metadata": {},
   "source": [
    "## First Need to download data using gdc-cleint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25667d17",
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist = glob.glob(os.path.join(os.getcwd(),\"data/*/*.tsv\"))\n",
    "filelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e7cb2bb4-d618-4bcc-9bc2-675524961225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>file_name</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>...</th>\n",
       "      <th>AL451106.1</th>\n",
       "      <th>AC092910.4</th>\n",
       "      <th>AC073611.1</th>\n",
       "      <th>AC136977.1</th>\n",
       "      <th>AC078856.1</th>\n",
       "      <th>AC008763.4</th>\n",
       "      <th>AL592295.6</th>\n",
       "      <th>AC006486.3</th>\n",
       "      <th>AL391628.1</th>\n",
       "      <th>AP006621.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e99de328-1950-4739-a59b-e0685c45b94b</td>\n",
       "      <td>29.7759</td>\n",
       "      <td>0.1208</td>\n",
       "      <td>49.7936</td>\n",
       "      <td>4.7416</td>\n",
       "      <td>2.6487</td>\n",
       "      <td>9.8434</td>\n",
       "      <td>13.0292</td>\n",
       "      <td>11.4386</td>\n",
       "      <td>20.3905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2921</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.1013</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0124</td>\n",
       "      <td>0.7077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>a6384a57-0d59-4e92-8ca2-07bd22cf906f</td>\n",
       "      <td>56.2901</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>83.4729</td>\n",
       "      <td>6.5013</td>\n",
       "      <td>12.3590</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>0.0068</td>\n",
       "      <td>27.7641</td>\n",
       "      <td>1.1546</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.8124</td>\n",
       "      <td>0.2671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5654</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.4693</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>1.4564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>933a9a9c-26e4-4209-86f3-f18b812bfbdc</td>\n",
       "      <td>66.3745</td>\n",
       "      <td>0.5016</td>\n",
       "      <td>75.2411</td>\n",
       "      <td>5.3781</td>\n",
       "      <td>3.8030</td>\n",
       "      <td>17.9505</td>\n",
       "      <td>1.9869</td>\n",
       "      <td>15.8284</td>\n",
       "      <td>74.2166</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3681</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.3922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0128</td>\n",
       "      <td>1.1752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7412a444-9fad-4d96-a4d4-2c270045919a</td>\n",
       "      <td>57.2746</td>\n",
       "      <td>0.5082</td>\n",
       "      <td>92.0853</td>\n",
       "      <td>5.6865</td>\n",
       "      <td>2.9126</td>\n",
       "      <td>18.4282</td>\n",
       "      <td>27.8222</td>\n",
       "      <td>52.6998</td>\n",
       "      <td>14.0285</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.4905</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0425</td>\n",
       "      <td>3.2065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>325f91c9-70b9-4285-a7ff-0b7b6280a8ed</td>\n",
       "      <td>128.4459</td>\n",
       "      <td>7.0286</td>\n",
       "      <td>84.6279</td>\n",
       "      <td>5.7158</td>\n",
       "      <td>5.8681</td>\n",
       "      <td>17.5790</td>\n",
       "      <td>10.0602</td>\n",
       "      <td>70.3848</td>\n",
       "      <td>21.2514</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3641</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.8325</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>2.7076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>f053a1a1-718a-46bd-a761-8ed3511aeb57</td>\n",
       "      <td>44.7003</td>\n",
       "      <td>0.0803</td>\n",
       "      <td>55.1775</td>\n",
       "      <td>5.8021</td>\n",
       "      <td>2.0247</td>\n",
       "      <td>4.6252</td>\n",
       "      <td>1.1732</td>\n",
       "      <td>4.7872</td>\n",
       "      <td>42.5586</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.6909</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0616</td>\n",
       "      <td>4.3028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>602acda6-a90d-42bc-a8aa-6eade2a791c4</td>\n",
       "      <td>47.4886</td>\n",
       "      <td>0.3089</td>\n",
       "      <td>66.6481</td>\n",
       "      <td>5.7443</td>\n",
       "      <td>4.0697</td>\n",
       "      <td>13.6360</td>\n",
       "      <td>9.4849</td>\n",
       "      <td>52.6728</td>\n",
       "      <td>14.5572</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.3318</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0135</td>\n",
       "      <td>2.9467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>660</th>\n",
       "      <td>4b228438-6c20-4c56-8633-e1873d1aee8c</td>\n",
       "      <td>43.6628</td>\n",
       "      <td>0.4114</td>\n",
       "      <td>65.7862</td>\n",
       "      <td>6.5181</td>\n",
       "      <td>2.7338</td>\n",
       "      <td>6.8011</td>\n",
       "      <td>2.2545</td>\n",
       "      <td>12.0243</td>\n",
       "      <td>36.5026</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5861</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.1121</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0710</td>\n",
       "      <td>2.8467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>04d8f7fc-783b-4c42-89cc-35d603174e5b</td>\n",
       "      <td>25.1427</td>\n",
       "      <td>0.3060</td>\n",
       "      <td>41.0510</td>\n",
       "      <td>7.8615</td>\n",
       "      <td>6.5292</td>\n",
       "      <td>6.5307</td>\n",
       "      <td>14.5350</td>\n",
       "      <td>16.8377</td>\n",
       "      <td>11.6456</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.2918</td>\n",
       "      <td>0.2615</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.7707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0990</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0657</td>\n",
       "      <td>1.2636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>715bd371-4f22-4f97-a9df-ebaf34715a1e</td>\n",
       "      <td>71.9935</td>\n",
       "      <td>0.2046</td>\n",
       "      <td>86.3821</td>\n",
       "      <td>8.5481</td>\n",
       "      <td>9.8149</td>\n",
       "      <td>3.2686</td>\n",
       "      <td>3.5427</td>\n",
       "      <td>26.9714</td>\n",
       "      <td>38.3417</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0215</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0126</td>\n",
       "      <td>4.3148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>663 rows × 60661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name                             file_name    TSPAN6    TNMD     DPM1  \\\n",
       "0          e99de328-1950-4739-a59b-e0685c45b94b   29.7759  0.1208  49.7936   \n",
       "1          a6384a57-0d59-4e92-8ca2-07bd22cf906f   56.2901  0.0737  83.4729   \n",
       "2          933a9a9c-26e4-4209-86f3-f18b812bfbdc   66.3745  0.5016  75.2411   \n",
       "3          7412a444-9fad-4d96-a4d4-2c270045919a   57.2746  0.5082  92.0853   \n",
       "4          325f91c9-70b9-4285-a7ff-0b7b6280a8ed  128.4459  7.0286  84.6279   \n",
       "..                                          ...       ...     ...      ...   \n",
       "658        f053a1a1-718a-46bd-a761-8ed3511aeb57   44.7003  0.0803  55.1775   \n",
       "659        602acda6-a90d-42bc-a8aa-6eade2a791c4   47.4886  0.3089  66.6481   \n",
       "660        4b228438-6c20-4c56-8633-e1873d1aee8c   43.6628  0.4114  65.7862   \n",
       "661        04d8f7fc-783b-4c42-89cc-35d603174e5b   25.1427  0.3060  41.0510   \n",
       "662        715bd371-4f22-4f97-a9df-ebaf34715a1e   71.9935  0.2046  86.3821   \n",
       "\n",
       "gene_name   SCYL3  C1orf112      FGR      CFH    FUCA2     GCLC  ...  \\\n",
       "0          4.7416    2.6487   9.8434  13.0292  11.4386  20.3905  ...   \n",
       "1          6.5013   12.3590   0.0322   0.0068  27.7641   1.1546  ...   \n",
       "2          5.3781    3.8030  17.9505   1.9869  15.8284  74.2166  ...   \n",
       "3          5.6865    2.9126  18.4282  27.8222  52.6998  14.0285  ...   \n",
       "4          5.7158    5.8681  17.5790  10.0602  70.3848  21.2514  ...   \n",
       "..            ...       ...      ...      ...      ...      ...  ...   \n",
       "658        5.8021    2.0247   4.6252   1.1732   4.7872  42.5586  ...   \n",
       "659        5.7443    4.0697  13.6360   9.4849  52.6728  14.5572  ...   \n",
       "660        6.5181    2.7338   6.8011   2.2545  12.0243  36.5026  ...   \n",
       "661        7.8615    6.5292   6.5307  14.5350  16.8377  11.6456  ...   \n",
       "662        8.5481    9.8149   3.2686   3.5427  26.9714  38.3417  ...   \n",
       "\n",
       "gene_name  AL451106.1  AC092910.4  AC073611.1  AC136977.1  AC078856.1  \\\n",
       "0                 0.0      0.0000      0.2921         0.0      0.0000   \n",
       "1                 0.0      1.8124      0.2671         0.0      3.5654   \n",
       "2                 0.0      0.0000      0.3681         0.0      0.0000   \n",
       "3                 0.0      0.0000      0.5025         0.0      0.0000   \n",
       "4                 0.0      0.0000      0.3641         0.0      0.0000   \n",
       "..                ...         ...         ...         ...         ...   \n",
       "658               0.0      0.0000      0.7070         0.0      0.0000   \n",
       "659               0.0      0.0000      0.2629         0.0      0.0000   \n",
       "660               0.0      0.0000      0.5861         0.0      0.0000   \n",
       "661               0.0     11.2918      0.2615         0.0     17.7707   \n",
       "662               0.0      0.0000      0.6254         0.0      0.0000   \n",
       "\n",
       "gene_name  AC008763.4  AL592295.6  AC006486.3  AL391628.1  AP006621.6  \n",
       "0                 0.0     25.1013         0.0      0.0124      0.7077  \n",
       "1                 0.0      6.4693         0.0      0.0226      1.4564  \n",
       "2                 0.0     26.3922         0.0      0.0128      1.1752  \n",
       "3                 0.0     10.4905         0.0      0.0425      3.2065  \n",
       "4                 0.0      8.8325         0.0      0.0166      2.7076  \n",
       "..                ...         ...         ...         ...         ...  \n",
       "658               0.0     22.6909         0.0      0.0616      4.3028  \n",
       "659               0.0     24.3318         0.0      0.0135      2.9467  \n",
       "660               0.0     34.1121         0.0      0.0710      2.8467  \n",
       "661               0.0      6.0990         0.0      0.0657      1.2636  \n",
       "662               0.0     19.0215         0.0      0.0126      4.3148  \n",
       "\n",
       "[663 rows x 60661 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the list of file paths\n",
    "# filelist = glob.glob(\"./data/*/*.tsv\")\n",
    "all_dfs = []\n",
    "\n",
    "for filepath in filelist:\n",
    "    filename =  os.path.basename(filepath)  # get filename from path\n",
    "    df = read_gene_df(filepath, filename)\n",
    "    all_dfs.append(df) \n",
    "\n",
    "combined_df = pd.concat(all_dfs, ignore_index=True)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3363615b-7e6c-4a2d-9b7a-4e50c45d9239",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CD99', 'NME1-NME2', 'Z83844.1', 'MCTS2P', 'CORO7-PAM16', 'ICAM4',\n",
       "       'AC004837.1', 'AL021546.1', 'PEDS1-UBE2V1', 'VAMP7',\n",
       "       ...\n",
       "       'CDR1', 'AC114982.2', 'AC114402.2', 'AC084756.2', 'AL031178.2',\n",
       "       'ACTL10', 'AC119733.1', 'AL451106.1', 'AC136977.1', 'AC006486.3'],\n",
       "      dtype='object', name='gene_name', length=3329)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sanity check for columns that have all zeros across rows (samples)\n",
    "zero_read_genes = combined_df.columns[(combined_df == 0).all(axis=0)]\n",
    "zero_read_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bee8a87b-340c-4f7a-afae-e38997116e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>file_name</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>...</th>\n",
       "      <th>AC010980.1</th>\n",
       "      <th>AC007511.1</th>\n",
       "      <th>AC004233.4</th>\n",
       "      <th>AC092910.4</th>\n",
       "      <th>AC073611.1</th>\n",
       "      <th>AC078856.1</th>\n",
       "      <th>AC008763.4</th>\n",
       "      <th>AL592295.6</th>\n",
       "      <th>AL391628.1</th>\n",
       "      <th>AP006621.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e99de328-1950-4739-a59b-e0685c45b94b</td>\n",
       "      <td>29.7759</td>\n",
       "      <td>0.1208</td>\n",
       "      <td>49.7936</td>\n",
       "      <td>4.7416</td>\n",
       "      <td>2.6487</td>\n",
       "      <td>9.8434</td>\n",
       "      <td>13.0292</td>\n",
       "      <td>11.4386</td>\n",
       "      <td>20.3905</td>\n",
       "      <td>...</td>\n",
       "      <td>1.3224</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2921</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>25.1013</td>\n",
       "      <td>0.0124</td>\n",
       "      <td>0.7077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>a6384a57-0d59-4e92-8ca2-07bd22cf906f</td>\n",
       "      <td>56.2901</td>\n",
       "      <td>0.0737</td>\n",
       "      <td>83.4729</td>\n",
       "      <td>6.5013</td>\n",
       "      <td>12.3590</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>0.0068</td>\n",
       "      <td>27.7641</td>\n",
       "      <td>1.1546</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1100</td>\n",
       "      <td>0.0928</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.8124</td>\n",
       "      <td>0.2671</td>\n",
       "      <td>3.5654</td>\n",
       "      <td>0.000</td>\n",
       "      <td>6.4693</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>1.4564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>99618995-9556-4c65-a643-43c9354395a7</td>\n",
       "      <td>59.3714</td>\n",
       "      <td>0.1389</td>\n",
       "      <td>43.7528</td>\n",
       "      <td>4.6948</td>\n",
       "      <td>1.5906</td>\n",
       "      <td>4.7471</td>\n",
       "      <td>1.9091</td>\n",
       "      <td>16.5347</td>\n",
       "      <td>41.5551</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2072</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1879</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5874</td>\n",
       "      <td>1.1200</td>\n",
       "      <td>0.000</td>\n",
       "      <td>23.7052</td>\n",
       "      <td>0.0568</td>\n",
       "      <td>1.9113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5b1cc683-d363-4cb8-b9b2-050c3c541858</td>\n",
       "      <td>30.2841</td>\n",
       "      <td>0.1136</td>\n",
       "      <td>92.6183</td>\n",
       "      <td>5.8728</td>\n",
       "      <td>4.1862</td>\n",
       "      <td>18.3498</td>\n",
       "      <td>1.6605</td>\n",
       "      <td>24.4281</td>\n",
       "      <td>15.9592</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3392</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2307</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.9572</td>\n",
       "      <td>1.3748</td>\n",
       "      <td>0.021</td>\n",
       "      <td>9.6998</td>\n",
       "      <td>0.0436</td>\n",
       "      <td>3.9436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>e7e63873-0099-474a-94b1-160a155c1ea3</td>\n",
       "      <td>16.8758</td>\n",
       "      <td>0.3241</td>\n",
       "      <td>56.2293</td>\n",
       "      <td>4.1407</td>\n",
       "      <td>1.1105</td>\n",
       "      <td>5.6585</td>\n",
       "      <td>2.1326</td>\n",
       "      <td>12.6668</td>\n",
       "      <td>22.7295</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0383</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0940</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4677</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>28.7837</td>\n",
       "      <td>0.0142</td>\n",
       "      <td>2.6037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>e11c6b6a-61cc-48ef-ac4d-221f27d45091</td>\n",
       "      <td>23.7059</td>\n",
       "      <td>0.4353</td>\n",
       "      <td>44.5254</td>\n",
       "      <td>3.2805</td>\n",
       "      <td>1.9987</td>\n",
       "      <td>3.5011</td>\n",
       "      <td>3.6117</td>\n",
       "      <td>8.3267</td>\n",
       "      <td>14.4904</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5414</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0631</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3221</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>15.9500</td>\n",
       "      <td>0.0429</td>\n",
       "      <td>1.0381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>843f0cc6-cd74-4af8-be28-5fb02b532269</td>\n",
       "      <td>89.1621</td>\n",
       "      <td>4.1974</td>\n",
       "      <td>56.5681</td>\n",
       "      <td>8.1008</td>\n",
       "      <td>4.5077</td>\n",
       "      <td>8.1289</td>\n",
       "      <td>19.8376</td>\n",
       "      <td>56.2106</td>\n",
       "      <td>26.5987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1812</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>30.1781</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>2.5353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>66cfd7f5-5c5e-4d20-b2b2-c3c592189160</td>\n",
       "      <td>59.8688</td>\n",
       "      <td>0.1680</td>\n",
       "      <td>62.5148</td>\n",
       "      <td>6.2605</td>\n",
       "      <td>1.8694</td>\n",
       "      <td>11.2746</td>\n",
       "      <td>3.9081</td>\n",
       "      <td>5.0972</td>\n",
       "      <td>26.1516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1463</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5113</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>29.6109</td>\n",
       "      <td>0.0064</td>\n",
       "      <td>3.1739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>e3c51e59-d9ff-425c-ade2-2a396333826a</td>\n",
       "      <td>33.4777</td>\n",
       "      <td>0.2491</td>\n",
       "      <td>53.7671</td>\n",
       "      <td>5.0215</td>\n",
       "      <td>6.3026</td>\n",
       "      <td>8.1720</td>\n",
       "      <td>7.9097</td>\n",
       "      <td>25.1466</td>\n",
       "      <td>13.9373</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1652</td>\n",
       "      <td>0.2092</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.1068</td>\n",
       "      <td>0.2473</td>\n",
       "      <td>20.0922</td>\n",
       "      <td>0.000</td>\n",
       "      <td>8.1392</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>1.8238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>a4e0b059-6e1e-4e90-b55a-4e043e6ed051</td>\n",
       "      <td>50.9506</td>\n",
       "      <td>0.1954</td>\n",
       "      <td>52.8074</td>\n",
       "      <td>4.9584</td>\n",
       "      <td>3.4783</td>\n",
       "      <td>16.4870</td>\n",
       "      <td>7.6500</td>\n",
       "      <td>49.2947</td>\n",
       "      <td>19.6222</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2637</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3880</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>13.1942</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>0.9728</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>222 rows × 56242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name                             file_name   TSPAN6    TNMD     DPM1  \\\n",
       "0          e99de328-1950-4739-a59b-e0685c45b94b  29.7759  0.1208  49.7936   \n",
       "1          a6384a57-0d59-4e92-8ca2-07bd22cf906f  56.2901  0.0737  83.4729   \n",
       "2          99618995-9556-4c65-a643-43c9354395a7  59.3714  0.1389  43.7528   \n",
       "3          5b1cc683-d363-4cb8-b9b2-050c3c541858  30.2841  0.1136  92.6183   \n",
       "4          e7e63873-0099-474a-94b1-160a155c1ea3  16.8758  0.3241  56.2293   \n",
       "..                                          ...      ...     ...      ...   \n",
       "217        e11c6b6a-61cc-48ef-ac4d-221f27d45091  23.7059  0.4353  44.5254   \n",
       "218        843f0cc6-cd74-4af8-be28-5fb02b532269  89.1621  4.1974  56.5681   \n",
       "219        66cfd7f5-5c5e-4d20-b2b2-c3c592189160  59.8688  0.1680  62.5148   \n",
       "220        e3c51e59-d9ff-425c-ade2-2a396333826a  33.4777  0.2491  53.7671   \n",
       "221        a4e0b059-6e1e-4e90-b55a-4e043e6ed051  50.9506  0.1954  52.8074   \n",
       "\n",
       "gene_name   SCYL3  C1orf112      FGR      CFH    FUCA2     GCLC  ...  \\\n",
       "0          4.7416    2.6487   9.8434  13.0292  11.4386  20.3905  ...   \n",
       "1          6.5013   12.3590   0.0322   0.0068  27.7641   1.1546  ...   \n",
       "2          4.6948    1.5906   4.7471   1.9091  16.5347  41.5551  ...   \n",
       "3          5.8728    4.1862  18.3498   1.6605  24.4281  15.9592  ...   \n",
       "4          4.1407    1.1105   5.6585   2.1326  12.6668  22.7295  ...   \n",
       "..            ...       ...      ...      ...      ...      ...  ...   \n",
       "217        3.2805    1.9987   3.5011   3.6117   8.3267  14.4904  ...   \n",
       "218        8.1008    4.5077   8.1289  19.8376  56.2106  26.5987  ...   \n",
       "219        6.2605    1.8694  11.2746   3.9081   5.0972  26.1516  ...   \n",
       "220        5.0215    6.3026   8.1720   7.9097  25.1466  13.9373  ...   \n",
       "221        4.9584    3.4783  16.4870   7.6500  49.2947  19.6222  ...   \n",
       "\n",
       "gene_name  AC010980.1  AC007511.1  AC004233.4  AC092910.4  AC073611.1  \\\n",
       "0              1.3224      0.0000      0.0000      0.0000      0.2921   \n",
       "1              0.1100      0.0928      0.0000      1.8124      0.2671   \n",
       "2              0.2072      0.0000      0.1879      0.0000      0.5874   \n",
       "3              0.3392      0.0000      0.2307      0.0000      1.9572   \n",
       "4             13.0383      0.0000      0.0940      0.0000      0.4677   \n",
       "..                ...         ...         ...         ...         ...   \n",
       "217            0.5414      0.0000      0.0631      0.0000      0.3221   \n",
       "218            0.8700      0.0000      0.0000      0.0000      0.1812   \n",
       "219            0.1463      0.0000      0.0000      0.0000      0.5113   \n",
       "220            0.1652      0.2092      0.0000      5.1068      0.2473   \n",
       "221            1.2637      0.0000      0.0000      0.0000      0.3880   \n",
       "\n",
       "gene_name  AC078856.1  AC008763.4  AL592295.6  AL391628.1  AP006621.6  \n",
       "0              0.0000       0.000     25.1013      0.0124      0.7077  \n",
       "1              3.5654       0.000      6.4693      0.0226      1.4564  \n",
       "2              1.1200       0.000     23.7052      0.0568      1.9113  \n",
       "3              1.3748       0.021      9.6998      0.0436      3.9436  \n",
       "4              0.0000       0.000     28.7837      0.0142      2.6037  \n",
       "..                ...         ...         ...         ...         ...  \n",
       "217            0.0000       0.000     15.9500      0.0429      1.0381  \n",
       "218            0.0000       0.000     30.1781      0.0268      2.5353  \n",
       "219            0.0000       0.000     29.6109      0.0064      3.1739  \n",
       "220           20.0922       0.000      8.1392      0.0637      1.8238  \n",
       "221            0.0000       0.000     13.1942      0.0300      0.9728  \n",
       "\n",
       "[222 rows x 56242 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop zero count genes\n",
    "combined_df = combined_df.drop(columns=zero_read_genes)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353d47c2-89fd-463d-ae2b-471338f7d7c3",
   "metadata": {},
   "source": [
    "## Read clinical and sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "141d1dec-6c6b-4927-ba9b-cd4e2659d46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>age_is_obfuscated</th>\n",
       "      <th>cause_of_death</th>\n",
       "      <th>cause_of_death_source</th>\n",
       "      <th>country_of_residence_at_enrollment</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_arm</th>\n",
       "      <th>treatment_dose</th>\n",
       "      <th>treatment_dose_units</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_effect_indicator</th>\n",
       "      <th>treatment_frequency</th>\n",
       "      <th>treatment_intent_type</th>\n",
       "      <th>treatment_or_therapy</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>treatment_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Immunotherapy (Including Vaccines)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Targeted Molecular Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Chemotherapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Neoadjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 158 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id  case_submitter_id project_id  \\\n",
       "0  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "1  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "2  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "3  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "4  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "\n",
       "  age_at_index age_is_obfuscated  cause_of_death cause_of_death_source  \\\n",
       "0          '--             False  Cancer Related                   '--   \n",
       "1          '--             False  Cancer Related                   '--   \n",
       "2          '--             False  Cancer Related                   '--   \n",
       "3          '--             False  Cancer Related                   '--   \n",
       "4          '--             False  Cancer Related                   '--   \n",
       "\n",
       "  country_of_residence_at_enrollment days_to_birth days_to_death  ...  \\\n",
       "0                                '--        -19586           481  ...   \n",
       "1                                '--        -19586           481  ...   \n",
       "2                                '--        -19586           481  ...   \n",
       "3                                '--        -19586           481  ...   \n",
       "4                                '--        -19586           481  ...   \n",
       "\n",
       "  treatment_arm treatment_dose treatment_dose_units treatment_effect  \\\n",
       "0           '--            '--                  '--              '--   \n",
       "1           '--            '--                  '--              '--   \n",
       "2           '--            '--                  '--              '--   \n",
       "3           '--            '--                  '--              '--   \n",
       "4           '--            '--                  '--              '--   \n",
       "\n",
       "  treatment_effect_indicator treatment_frequency treatment_intent_type  \\\n",
       "0                        '--                 '--              Adjuvant   \n",
       "1                        '--                 '--              Adjuvant   \n",
       "2                        '--                 '--              Adjuvant   \n",
       "3                        '--                 '--              Adjuvant   \n",
       "4                        '--                 '--           Neoadjuvant   \n",
       "\n",
       "   treatment_or_therapy treatment_outcome                      treatment_type  \n",
       "0                    no               '--  Immunotherapy (Including Vaccines)  \n",
       "1                    no               '--          Targeted Molecular Therapy  \n",
       "2                    no               '--              Radiation Therapy, NOS  \n",
       "3                   yes               '--                        Chemotherapy  \n",
       "4                    no               '--                                 '--  \n",
       "\n",
       "[5 rows x 158 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load clinical data \n",
    "clinical_data = pd.read_csv(\"/home/talal/Documents/Projects/fyp/Glioma-ML-Classifier-with-ANOVA-Feature-Selection/metadata/clinical.tsv\", sep='\\t')\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8ee7956e-5102-463a-b969-457267b78123",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1386, 158)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cd294ea7-0118-4f98-8acd-fe4e40cef8f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "primary_diagnosis\n",
       "Astrocytoma, NOS                 118\n",
       "Astrocytoma, anaplastic          240\n",
       "Glioblastoma                     434\n",
       "Gliosarcoma                        6\n",
       "Mixed glioma                     242\n",
       "Oligodendroglioma, NOS           202\n",
       "Oligodendroglioma, anaplastic    144\n",
       "Name: case_id, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.groupby(\"primary_diagnosis\")['case_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9328141a-0624-4d93-b904-0721dfc7dc35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(663,\n",
       " case_submitter_id\n",
       " HCM-BROD-0213-C71    8\n",
       " HCM-BROD-0198-C71    7\n",
       " HCM-BROD-0106-C71    7\n",
       " HCM-BROD-0104-C71    7\n",
       " HCM-BROD-0420-C71    7\n",
       " Name: count, dtype: int64)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_cases_in_clinical = clinical_data['case_submitter_id'].nunique()\n",
    "case_duplicates = clinical_data['case_submitter_id'].value_counts()\n",
    "# Display the unique case count and the distribution of duplicates\n",
    "unique_cases_in_clinical, case_duplicates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "450596e9-f792-402d-bc78-dc3ae9ae44c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(663, 158)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicates based on 'case_submitter_id'\n",
    "clinical_data_dedup = clinical_data.drop_duplicates(subset='case_submitter_id', keep='first')\n",
    "clinical_data_dedup.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c257052e-2716-483e-847a-92535e090890",
   "metadata": {},
   "source": [
    "## Simplify labels to broader categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "37a0b644-2fba-4fb9-8a5c-30d360ca782d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_32651/80191726.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinical_data_dedup.loc[:, 'label'] = clinical_data_dedup['primary_diagnosis'].map(diagnosis_map)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "primary_diagnosis              label            \n",
       "Astrocytoma, NOS               Astrocytoma           59\n",
       "Astrocytoma, anaplastic        Astrocytoma          120\n",
       "Glioblastoma                   Glioblastoma         189\n",
       "Oligodendroglioma, NOS         Oligodendroglioma    101\n",
       "Oligodendroglioma, anaplastic  Oligodendroglioma     72\n",
       "Name: case_submitter_id, dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# simplify labels by mapping complex or specific diagnoses to broader categories - dropping 'mixed gliomas'\n",
    "diagnosis_map = {\n",
    "    \"Astrocytoma, NOS\": \"Astrocytoma\",\n",
    "    \"Astrocytoma, anaplastic\": \"Astrocytoma\",\n",
    "    \"Oligodendroglioma, NOS\": \"Oligodendroglioma\",\n",
    "    \"Oligodendroglioma, anaplastic\": \"Oligodendroglioma\",\n",
    "    \"Glioblastoma\": \"Glioblastoma\"\n",
    "}\n",
    "\n",
    "# create a new 'label' column based on the mapping\n",
    "clinical_data_dedup.loc[:, 'label'] = clinical_data_dedup['primary_diagnosis'].map(diagnosis_map)\n",
    "clinical_data_dedup.groupby([\"primary_diagnosis\", \"label\"])['case_submitter_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "97f7826e-6b31-4c2d-8fdb-abce4e0824c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>filename_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4fa35d6c-060b-433b-bc26-8bd7984631cc</td>\n",
       "      <td>9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCM-BROD-0210-C71-85R</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d9c44441-5f47-4532-8ac1-8a0a7a59a6f0</td>\n",
       "      <td>7835cada-4a04-42d0-bed4-e8af440b5913.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0209-C71</td>\n",
       "      <td>HCM-BROD-0209-C71-85A</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>7835cada-4a04-42d0-bed4-e8af440b5913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6807181b-13e7-4f4b-a7d8-9d00f113f6bc</td>\n",
       "      <td>517e37c3-64f8-40fa-8b6c-3c74e423ea1a.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03266</td>\n",
       "      <td>C3L-03266-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>517e37c3-64f8-40fa-8b6c-3c74e423ea1a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f23b7ee8-2af9-4db6-9f2e-bbb74ce501c7</td>\n",
       "      <td>c8b55027-80ef-4748-940a-1c97e0e12116.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03727</td>\n",
       "      <td>C3L-03727-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>c8b55027-80ef-4748-940a-1c97e0e12116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>849a5e37-7da5-4d58-9790-9868792456b0</td>\n",
       "      <td>5b664467-3c27-4bf4-90f8-4f518a153dd8.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-01887</td>\n",
       "      <td>C3L-01887-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5b664467-3c27-4bf4-90f8-4f518a153dd8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  4fa35d6c-060b-433b-bc26-8bd7984631cc   \n",
       "1  d9c44441-5f47-4532-8ac1-8a0a7a59a6f0   \n",
       "2  6807181b-13e7-4f4b-a7d8-9d00f113f6bc   \n",
       "3  f23b7ee8-2af9-4db6-9f2e-bbb74ce501c7   \n",
       "4  849a5e37-7da5-4d58-9790-9868792456b0   \n",
       "\n",
       "                                           File Name            Data Category  \\\n",
       "0  9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d.rna_seq.a...  Transcriptome Profiling   \n",
       "1  7835cada-4a04-42d0-bed4-e8af440b5913.rna_seq.a...  Transcriptome Profiling   \n",
       "2  517e37c3-64f8-40fa-8b6c-3c74e423ea1a.rna_seq.a...  Transcriptome Profiling   \n",
       "3  c8b55027-80ef-4748-940a-1c97e0e12116.rna_seq.a...  Transcriptome Profiling   \n",
       "4  5b664467-3c27-4bf4-90f8-4f518a153dd8.rna_seq.a...  Transcriptome Profiling   \n",
       "\n",
       "                        Data Type Project ID            Case ID  \\\n",
       "0  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0210-C71   \n",
       "1  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0209-C71   \n",
       "2  Gene Expression Quantification    CPTAC-3          C3L-03266   \n",
       "3  Gene Expression Quantification    CPTAC-3          C3L-03727   \n",
       "4  Gene Expression Quantification    CPTAC-3          C3L-01887   \n",
       "\n",
       "               Sample ID                   Sample Type  \\\n",
       "0  HCM-BROD-0210-C71-85R  Next Generation Cancer Model   \n",
       "1  HCM-BROD-0209-C71-85A  Next Generation Cancer Model   \n",
       "2           C3L-03266-01                 Primary Tumor   \n",
       "3           C3L-03727-01                 Primary Tumor   \n",
       "4           C3L-01887-01                 Primary Tumor   \n",
       "\n",
       "                         filename_short  \n",
       "0  9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d  \n",
       "1  7835cada-4a04-42d0-bed4-e8af440b5913  \n",
       "2  517e37c3-64f8-40fa-8b6c-3c74e423ea1a  \n",
       "3  c8b55027-80ef-4748-940a-1c97e0e12116  \n",
       "4  5b664467-3c27-4bf4-90f8-4f518a153dd8  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_data = pd.read_csv(\"/home/talal/Documents/Projects/fyp/Glioma-ML-Classifier-with-ANOVA-Feature-Selection/metadata/gdc_sample_sheet.tsv\", sep='\\t')\n",
    "# Simplify filenames in sample_data for merging\n",
    "sample_data['filename_short'] = sample_data['File Name'].apply(lambda x: x.split('.')[0])\n",
    "sample_data = pd.merge(sample_data, combined_df, left_on='filename_short', right_on='file_name', how='inner')\n",
    "\n",
    "# Retain relevant columns in sample_data\n",
    "columns_to_retain = ['File ID', 'File Name', 'Data Category', 'Data Type', 'Project ID', 'Case ID', 'Sample ID', 'Sample Type', 'filename_short']\n",
    "sample_data = sample_data[columns_to_retain]\n",
    "sample_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a4342ff7-b5f7-4042-a734-ad4714fda691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(707, 9)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "37c31ebc-6d67-42e6-8c0e-84977263294a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>filename_short</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4fa35d6c-060b-433b-bc26-8bd7984631cc</td>\n",
       "      <td>9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCM-BROD-0210-C71-85R</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d9c44441-5f47-4532-8ac1-8a0a7a59a6f0</td>\n",
       "      <td>7835cada-4a04-42d0-bed4-e8af440b5913.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0209-C71</td>\n",
       "      <td>HCM-BROD-0209-C71-85A</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>7835cada-4a04-42d0-bed4-e8af440b5913</td>\n",
       "      <td>HCM-BROD-0209-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6807181b-13e7-4f4b-a7d8-9d00f113f6bc</td>\n",
       "      <td>517e37c3-64f8-40fa-8b6c-3c74e423ea1a.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03266</td>\n",
       "      <td>C3L-03266-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>517e37c3-64f8-40fa-8b6c-3c74e423ea1a</td>\n",
       "      <td>C3L-03266</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>f23b7ee8-2af9-4db6-9f2e-bbb74ce501c7</td>\n",
       "      <td>c8b55027-80ef-4748-940a-1c97e0e12116.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03727</td>\n",
       "      <td>C3L-03727-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>c8b55027-80ef-4748-940a-1c97e0e12116</td>\n",
       "      <td>C3L-03727</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>849a5e37-7da5-4d58-9790-9868792456b0</td>\n",
       "      <td>5b664467-3c27-4bf4-90f8-4f518a153dd8.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-01887</td>\n",
       "      <td>C3L-01887-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5b664467-3c27-4bf4-90f8-4f518a153dd8</td>\n",
       "      <td>C3L-01887</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>2761c4d4-00a9-4eb5-9166-af4e56677d07</td>\n",
       "      <td>a93d22ba-5cfa-4359-ae53-3309ced4e12a.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-02041</td>\n",
       "      <td>C3L-02041-02</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>a93d22ba-5cfa-4359-ae53-3309ced4e12a</td>\n",
       "      <td>C3L-02041</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>0997ec63-6fe0-40fb-b0ff-3ddcb14ec7d2</td>\n",
       "      <td>b7effa07-1b05-412a-a1ef-54d52d795c90.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-00365</td>\n",
       "      <td>C3L-00365-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>b7effa07-1b05-412a-a1ef-54d52d795c90</td>\n",
       "      <td>C3L-00365</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>2c259bc9-5e6a-46c0-b82b-d6f6eb257455</td>\n",
       "      <td>0dbfa70d-2a0d-41c4-aca6-b8df6394bb9c.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-02542</td>\n",
       "      <td>C3L-02542-03</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>0dbfa70d-2a0d-41c4-aca6-b8df6394bb9c</td>\n",
       "      <td>C3L-02542</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>3ae2f730-7444-4b65-8c75-1f8f18e6047a</td>\n",
       "      <td>27ad3b30-97ff-44a4-89f7-5233a3c3596e.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-02970</td>\n",
       "      <td>C3L-02970-03</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>27ad3b30-97ff-44a4-89f7-5233a3c3596e</td>\n",
       "      <td>C3L-02970</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>73d57ab1-1757-4af3-b229-90050dcdb9e0</td>\n",
       "      <td>1fbd55b4-c08b-44b8-940e-6538cc79ef68.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>C3L-00104-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1fbd55b4-c08b-44b8-940e-6538cc79ef68</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>222 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  File ID  \\\n",
       "0    4fa35d6c-060b-433b-bc26-8bd7984631cc   \n",
       "1    d9c44441-5f47-4532-8ac1-8a0a7a59a6f0   \n",
       "2    6807181b-13e7-4f4b-a7d8-9d00f113f6bc   \n",
       "3    f23b7ee8-2af9-4db6-9f2e-bbb74ce501c7   \n",
       "4    849a5e37-7da5-4d58-9790-9868792456b0   \n",
       "..                                    ...   \n",
       "217  2761c4d4-00a9-4eb5-9166-af4e56677d07   \n",
       "218  0997ec63-6fe0-40fb-b0ff-3ddcb14ec7d2   \n",
       "219  2c259bc9-5e6a-46c0-b82b-d6f6eb257455   \n",
       "220  3ae2f730-7444-4b65-8c75-1f8f18e6047a   \n",
       "221  73d57ab1-1757-4af3-b229-90050dcdb9e0   \n",
       "\n",
       "                                             File Name  \\\n",
       "0    9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d.rna_seq.a...   \n",
       "1    7835cada-4a04-42d0-bed4-e8af440b5913.rna_seq.a...   \n",
       "2    517e37c3-64f8-40fa-8b6c-3c74e423ea1a.rna_seq.a...   \n",
       "3    c8b55027-80ef-4748-940a-1c97e0e12116.rna_seq.a...   \n",
       "4    5b664467-3c27-4bf4-90f8-4f518a153dd8.rna_seq.a...   \n",
       "..                                                 ...   \n",
       "217  a93d22ba-5cfa-4359-ae53-3309ced4e12a.rna_seq.a...   \n",
       "218  b7effa07-1b05-412a-a1ef-54d52d795c90.rna_seq.a...   \n",
       "219  0dbfa70d-2a0d-41c4-aca6-b8df6394bb9c.rna_seq.a...   \n",
       "220  27ad3b30-97ff-44a4-89f7-5233a3c3596e.rna_seq.a...   \n",
       "221  1fbd55b4-c08b-44b8-940e-6538cc79ef68.rna_seq.a...   \n",
       "\n",
       "               Data Category                       Data Type Project ID  \\\n",
       "0    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "1    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "2    Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "3    Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "4    Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "..                       ...                             ...        ...   \n",
       "217  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "218  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "219  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "220  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "221  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "\n",
       "               Case ID              Sample ID                   Sample Type  \\\n",
       "0    HCM-BROD-0210-C71  HCM-BROD-0210-C71-85R  Next Generation Cancer Model   \n",
       "1    HCM-BROD-0209-C71  HCM-BROD-0209-C71-85A  Next Generation Cancer Model   \n",
       "2            C3L-03266           C3L-03266-01                 Primary Tumor   \n",
       "3            C3L-03727           C3L-03727-01                 Primary Tumor   \n",
       "4            C3L-01887           C3L-01887-01                 Primary Tumor   \n",
       "..                 ...                    ...                           ...   \n",
       "217          C3L-02041           C3L-02041-02                 Primary Tumor   \n",
       "218          C3L-00365           C3L-00365-01                 Primary Tumor   \n",
       "219          C3L-02542           C3L-02542-03                 Primary Tumor   \n",
       "220          C3L-02970           C3L-02970-03                 Primary Tumor   \n",
       "221          C3L-00104           C3L-00104-01                 Primary Tumor   \n",
       "\n",
       "                           filename_short  case_submitter_id  \\\n",
       "0    9e60d3ee-fcb2-48dc-8f3a-3171d25bf32d  HCM-BROD-0210-C71   \n",
       "1    7835cada-4a04-42d0-bed4-e8af440b5913  HCM-BROD-0209-C71   \n",
       "2    517e37c3-64f8-40fa-8b6c-3c74e423ea1a          C3L-03266   \n",
       "3    c8b55027-80ef-4748-940a-1c97e0e12116          C3L-03727   \n",
       "4    5b664467-3c27-4bf4-90f8-4f518a153dd8          C3L-01887   \n",
       "..                                    ...                ...   \n",
       "217  a93d22ba-5cfa-4359-ae53-3309ced4e12a          C3L-02041   \n",
       "218  b7effa07-1b05-412a-a1ef-54d52d795c90          C3L-00365   \n",
       "219  0dbfa70d-2a0d-41c4-aca6-b8df6394bb9c          C3L-02542   \n",
       "220  27ad3b30-97ff-44a4-89f7-5233a3c3596e          C3L-02970   \n",
       "221  1fbd55b4-c08b-44b8-940e-6538cc79ef68          C3L-00104   \n",
       "\n",
       "    primary_diagnosis         label  \n",
       "0        Glioblastoma  Glioblastoma  \n",
       "1        Glioblastoma  Glioblastoma  \n",
       "2        Glioblastoma  Glioblastoma  \n",
       "3        Glioblastoma  Glioblastoma  \n",
       "4        Glioblastoma  Glioblastoma  \n",
       "..                ...           ...  \n",
       "217      Glioblastoma  Glioblastoma  \n",
       "218      Glioblastoma  Glioblastoma  \n",
       "219      Glioblastoma  Glioblastoma  \n",
       "220      Glioblastoma  Glioblastoma  \n",
       "221      Glioblastoma  Glioblastoma  \n",
       "\n",
       "[222 rows x 12 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with the clinical data to get the primary diagnosis and label\n",
    "merged_data_dedup = sample_data.merge(\n",
    "    clinical_data_dedup[['case_submitter_id', 'primary_diagnosis', 'label']], \n",
    "    how='left', \n",
    "    left_on='Case ID', \n",
    "    right_on='case_submitter_id'\n",
    ")\n",
    "\n",
    "merged_data_dedup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "17f777f0-6ffb-44fb-a0d0-b77cce82a0c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(222, 12)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data_dedup.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fa87b0a8-553a-4073-a0c5-ea1eea06caf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "primary_diagnosis              label            \n",
       "Astrocytoma, NOS               Astrocytoma          17\n",
       "Astrocytoma, anaplastic        Astrocytoma          41\n",
       "Glioblastoma                   Glioblastoma         72\n",
       "Oligodendroglioma, NOS         Oligodendroglioma    29\n",
       "Oligodendroglioma, anaplastic  Oligodendroglioma    25\n",
       "Name: case_submitter_id, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by both primary_diagnosis and label, then count unique case_submitter_ids\n",
    "merged_data_dedup.groupby(['primary_diagnosis', 'label'])['case_submitter_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9a2fbf5c-bf9f-4ec2-9962-8329ebc34f92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(184, 12)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter out samples with non-empty labels\n",
    "merged_data_dedup = merged_data_dedup[merged_data_dedup['label'].notna()]\n",
    "merged_data_dedup.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "489564ba-c065-4acd-90f7-acb2b4a10bbd",
   "metadata": {},
   "source": [
    "## Keep samples that have valid labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1c507332-a705-405a-9b38-15b51cb9e94f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(184, 56241)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge combined_df with merged_data_dedup to get 'Sample ID'\n",
    "combined_df_with_sample_id = pd.merge(combined_df, merged_data_dedup[['filename_short', 'Sample ID']], \n",
    "    left_on='file_name', right_on='filename_short', how='left')\n",
    "\n",
    "# Set 'Sample ID' as index, drop unnecessary columns\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.set_index('Sample ID')\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.drop(['filename_short', 'file_name'], axis=1)\n",
    "\n",
    "# Filter combined_df_with_sample_id to only include non-empty labels\n",
    "filtered_sample_ids = merged_data_dedup['Sample ID'].dropna().unique()\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.loc[filtered_sample_ids]\n",
    "\n",
    "# Display the final shape\n",
    "combined_df_with_sample_id.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57eceaa5-dfff-46b9-ab74-a3cc297d6db4",
   "metadata": {},
   "source": [
    "## Output the Processed Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7c3fd380-374d-4e49-935b-391bebf49bff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved to ./processed/data.csv\n"
     ]
    }
   ],
   "source": [
    "# Define extraction directory\n",
    "extraction_dir = \"./processed\"\n",
    "os.makedirs(extraction_dir, exist_ok=True)\n",
    "\n",
    "data_file_path = os.path.join(extraction_dir, 'data.csv')\n",
    "combined_df_with_sample_id.to_csv(data_file_path, index=True)\n",
    "print(f\"Data saved to {data_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f88a9f5f-8429-4d0a-bc58-0378c38697a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Labels saved to ./processed/glioma_labels.csv\n"
     ]
    }
   ],
   "source": [
    "# Extract and save the filtered labels\n",
    "y = merged_data_dedup.set_index('Sample ID')['label']\n",
    "labels_file_path = os.path.join(extraction_dir, 'glioma_labels.csv')\n",
    "y.to_csv(labels_file_path, index=True)\n",
    "print(f\"Labels saved to {labels_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3537aeb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
